src/serial.rs

```diff
@@ -383,18 +465,43 @@ impl<W: Write> Serial<W> {
     /// # Example
     ///
     /// ```rust
+    /// # use std::io::{Error, Result};
+    /// # use std::ops::Deref;
+    /// # use vm_superio::trigger::Trigger;
     /// # use vm_superio::Serial;
     /// # use vmm_sys_util::eventfd::EventFd;
-    /// let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+    ///
+    /// struct TestEventFd(EventFd);
+    /// impl Trigger for TestEventFd {
+    ///     type E = Error;
+    ///
+    ///     fn trigger(&self) -> Result<()> {
+    ///         self.write(1)
+    ///     }
+    /// }
+    /// impl Deref for TestEventFd {
+    ///     type Target = EventFd;
+    ///     fn deref(&self) -> &Self::Target {
+    ///         &self.0
+    ///     }
+    /// }
+    /// impl TestEventFd {
+    ///     pub fn new(flag: i32) -> Self {
+    ///         TestEventFd(EventFd::new(flag).unwrap())
+    ///     }
+    /// }
+    ///
+    /// let intr_evt = TestEventFd::new(libc::EFD_NONBLOCK);
     /// let mut serial = Serial::new(intr_evt, Vec::new());
     ///
     /// serial.enqueue_raw_bytes(&[b'a', b'b', b'c']).unwrap();
     /// ```
-    pub fn enqueue_raw_bytes(&mut self, input: &[u8]) -> Result<()> {
+    pub fn enqueue_raw_bytes(&mut self, input: &[u8]) -> std::result::Result<(), Error<T::E>> {
         if !self.is_in_loop_mode() {
             self.in_buffer.extend(input);
             self.set_lsr_rda_bit();
-            self.received_data_interrupt()?;
+            self.received_data_interrupt()
+                .map_err(Error::TriggerInterrupt)?;
```

Comment:
How about we implement `From` for `Error` so that we do not need to call `map_err`?

Reply:
This turned out to be a bit tricky, since `Error` enum has the generic `E` type, which can be `io::Error` too (the same as the other variant's parameter). Also, the fact that we are returning here `Trigger::E` which is less generic than the other `E` seemed to make things weird. As for the i8042, I removed, at Alex's suggestion, the Error enum for now.

If you think it makes sense the effort to come up with a `From` impl here or some other errors improvement, we could log an issue, what do you say?